ocaml/xapi/repository.ml

```diff
@@ -149,26 +149,69 @@ let get_proxy_params ~__context repo_name =
   | _ ->
       ("", "", "")
 
-let sync ~__context ~self ~token ~token_id =
+type client_auth_conf =
+  | CdnTokenAuthConf (* remote *) of {token: string; token_id: string}
+  | NoAuth (* bundle *)
+  | PoolExtHostAuthConf (* remote_pool *) of {
+        cert: string
+      ; remote_addr: string
+      ; username: string
+      ; password: string
+    }
+
+type server_auth_conf =
+  | DefaultAuth (* remote *)
+  | NoAuth (* bundle *)
+  | ServerAuthConf (* remote_pool *) of {
+        cert: string
+      ; remote_addr: string
+      ; remote_port: int
+    }
+
+let sync ~__context ~self ~token ~token_id ~username ~password =
   try
     let repo_name = get_remote_repository_name ~__context ~self in
     remove_repo_conf_file repo_name ;
-    let binary_url, source_url =
-      match Db.Repository.get_origin ~__context ~self with
+    let origin = Db.Repository.get_origin ~__context ~self in
+    let local_host = "127.0.0.1" in
+    let local_port = !Xapi_globs.local_yum_repo_port in
+
+    let ( binary_url
+        , source_url
+        , repo_gpgcheck
+        , client_auth_config
+        , server_auth_config ) =
+      match origin with
       | `remote ->
           ( Db.Repository.get_binary_url ~__context ~self
           , Some (Db.Repository.get_source_url ~__context ~self)
+          , true
+          , CdnTokenAuthConf {token; token_id}
+          , DefaultAuth
           )
       | `bundle ->
           let uri =
             Uri.make ~scheme:"file" ~path:!Xapi_globs.bundle_repository_dir ()
           in
-          (Uri.to_string uri, None)
+          (Uri.to_string uri, None, true, NoAuth, NoAuth)
       | `remote_pool ->
-          (* TODO: sync with Stunnel.with_client_proxy as otherwise yum
-             reposync will fail when checking the self signed certificate on
-             the remote pool. *)
-          ("", None)
+          let uri =
+            Uri.make ~scheme:"http" ~host:local_host
```

Comment:
I think it's too early exposing the localhost here.

Reply:
It is used to config yum (line 224) just follows this.

Reply:
The `binary_url` is determined by the `ServerAuthConf`. It should be determined by the handling of `ServerAuthConf`.